src/python-common/ceph/deployment/translate.py

```diff
@@ -24,73 +28,146 @@ def __init__(self,
+            the drives together
+            """
+            if crush_device_class is None and self.spec.crush_device_class:
+                lvcount[self.spec.crush_device_class].append(path)
+                continue
+            else:
+                # default use case
+                lvcount["no_crush"].append(path)
+                continue
+
+        return lvcount
+
     def run(self):
         # type: () -> List[str]
         """ Generate ceph-volume commands based on the DriveGroup filters """
-        data_devices = [x.path for x in self.selection.data_devices()]
+
         db_devices = [x.path for x in self.selection.db_devices()]
         wal_devices = [x.path for x in self.selection.wal_devices()]
         journal_devices = [x.path for x in self.selection.journal_devices()]
 
-        if not data_devices:
+        if not self.selection.data_devices():
             return []
 
         cmds: List[str] = []
-        if self.spec.method == 'raw':
-            assert self.spec.objectstore == 'bluestore'
-            # ceph-volume raw prepare only support 1:1 ratio of data to db/wal devices
-            if data_devices and db_devices:
-                if len(data_devices) != len(db_devices):
-                    raise ValueError('Number of data devices must match number of '
-                                     'db devices for raw mode osds')
-            if data_devices and wal_devices:
-                if len(data_devices) != len(wal_devices):
-                    raise ValueError('Number of data devices must match number of '
-                                     'wal devices for raw mode osds')
-            # for raw prepare each data device needs its own prepare command
-            dev_counter = 0
-            while dev_counter < len(data_devices):
-                cmd = "raw prepare --bluestore"
-                cmd += " --data {}".format(data_devices[dev_counter])
-                if db_devices:
-                    cmd += " --block.db {}".format(db_devices[dev_counter])
-                if wal_devices:
-                    cmd += " --block.wal {}".format(wal_devices[dev_counter])
-                cmds.append(cmd)
-                dev_counter += 1
 
-        elif self.spec.objectstore == 'filestore':
-            # for lvm batch we can just do all devices in one command
-            cmd = "lvm batch --no-auto"
-
-            cmd += " {}".format(" ".join(data_devices))
-
-            if self.spec.journal_size:
-                cmd += " --journal-size {}".format(self.spec.journal_size)
+        devices = self.prepare_devices()
+
+        for d in devices.keys():
+            data_devices = devices.get(d)
+            if not data_devices:
+                continue
+
+            if self.spec.method == 'raw':
+                assert self.spec.objectstore == 'bluestore'
+                # ceph-volume raw prepare only support 1:1 ratio of data to db/wal devices
+                if data_devices and db_devices:
```

Comment:
> We probably need to be careful with this check. Previously `data_devices` was all the data devices provided. Now, it's just the data devices that match the `crush_device_class` we're currently iterating on. I don't think we need to raise an error here if one data device is hdd, one is ssd and there are 2 db devices (as this should be correct, 2 data and 2 db device == 1:1 ratio raw mode requires). It seems like we maybe want this check outside the for loop just checking total data devices against total db/wal devices. That does also bring up the tricky aspect that if we want to support device specific crush device classes for raw mode and we did have two data devices of different classes and 2 db devices, you would need to include one db device in the first command and the other in the second to maintain the 1:1 ratio in the `raw prepare` command(s) generated at the end. Or, if you don't want to implement that, we could maybe just move raw mode stuff out of this for loop entirely.

I see your point here, and it makes sense to see if it's easier remove this feature from the `raw prepare` phase and restrict to the `lvm` use case only. One of the goals of this patch is to avoid
breaking compatibility with the previous implementation, so I'll try to see if I can update the change for both `raw` and `filestore` use cases and provide a `distribute_db_wal` function (or something like that) that keep the 1:1 ratio between `data_devices` and `db/wal devices`.
I think the check on L115 can stay here, just because we have the portion of code:

```
        devices = self.prepare_devices()

        for d in devices.keys():
            data_devices = devices.get(d)
            if not data_devices:
                continue
            ....
            ....
```
that returns the `data_devices` from the new function instead of just checking the `path`, but the if statement should probably be different and related to a function where the ratio is computed and the provided `db/wal` devices are distributed to the commands.
I'll think a bit more about it but we can sync up if you have better ideas.


> It looks like currently it isn't actually using the device class `d` anyway so specifying it wouldn't have an effect in raw mode regardless. We could only support device specific crush device classes for the more popular lvm, bluestore case. I'd be okay with that as well as long as it gets included in the documentation.

`d` is actually used at the beginning because for each key of the dictionary (`no_crush`, `hdd`, `ssd`, etc) we get the corresponding `Devices` and we build the `ceph-volume` related command, but I agree this is the easiest way to solve the problem.
However, I'd like to see if there's a more generic way to apply the right behavior even when `crush_device_classes` are specified.